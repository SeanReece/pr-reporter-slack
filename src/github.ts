import * as core from '@actions/core'
import * as github from '@actions/github'

export enum ReviewStates {
  CHANGES_REQUESTED = 'CHANGES_REQUESTED',
  APPROVED = 'APPROVED',
}

export interface GraphQLResponse {
  nameWithOwner: string
  pullRequests: {
    nodes: PullRequest[]
    totalCount: number
  }
}

export interface PullRequest {
  id: string
  title: string
  url: string
  createdAt: string
  isDraft: boolean
  reviews: {
    totalCount: number
    nodes: [
      {
        state: keyof typeof ReviewStates
      },
    ]
  }
  comments: {
    totalCount: number
  }
  headRef: {
    name: string
  }
  commits: {
    nodes: object[]
  }
}

export async function queryPRs(token: string): Promise<GraphQLResponse> {
  try {
    const octokit = new github.GitHub(token)

    const response = await octokit.graphql(
      `query prs($owner: String!, $repo: String!) {
        repository(owner:$owner, name:$repo) {
          nameWithOwner,
          pullRequests(first: 100, states: OPEN) {
            nodes {
              id
              title
              url
              createdAt
              isDraft
              reviews(first: 10, states: [CHANGES_REQUESTED, APPROVED]) {
                totalCount
                nodes {
                  state
                }
              }
              comments {
                totalCount
              }
              headRef {
                name
              }
              commits(first: 10) {
                nodes {
                  commit {
                    status {
                      id
                      state
                    }
                  }
                }
              }
            }
            totalCount
          }
        }
      }`,
      {
        ...github.context.repo,
        headers: {
          accept: `application/vnd.github.shadow-cat-preview+json`,
        },
      },
    )

    return response && response.repository
  } catch (err) {
    core.error(`Could not perform github query: ${err.message}`)
    throw err
  }
}
