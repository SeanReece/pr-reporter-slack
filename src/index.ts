import * as github from '@actions/github'
import * as core from '@actions/core'
import axios from 'axios'
import { format } from 'timeago.js';

async function run(): Promise<void> {
  try {
    const token: string = core.getInput('repo-token')
    const slackWebhook: string = core.getInput('slack-webhook')

    const octokit = new github.GitHub(token)

    const response = await octokit.graphql(`query prs($owner: String!, $repo: String!) {
      repository(owner:$owner, name:$repo) {
        nameWithOwner,
        pullRequests(first: 100, states: OPEN) {
          nodes {
            id
            title
            url
            createdAt
            isDraft
            reviews(first: 10, states: [CHANGES_REQUESTED, APPROVED]) {
              totalCount
              nodes {
                state
              }
            }
            comments {
              totalCount
            }
            headRef {
              name
            }
            commits(first: 10) {
              nodes {
                commit {
                  status {
                    id
                    state
                  }
                }
              }
            }
          }
          totalCount
        }
      }
    }`, {
        ...github.context.repo, 
        headers: {
          accept: `application/vnd.github.shadow-cat-preview+json`
        }
      }
    )

    const pullRequests = response && response.repository.pullRequests.nodes
    const repoName = response && response.repository.nameWithOwner

    const readyPRS = pullRequests.filter((pr: any) => !pr.isDraft && !pr.title.toLowerCase().startsWith('[wip]'))

    let text = ''

    readyPRS.forEach((pr: any) => {
      let status = ''
      if (pr.reviews.totalCount === 0) {
        status = '*No reviews*'
      } else if (pr.reviews.nodes.some((review: any) => review.state === 'CHANGES_REQUESTED')) {
        status = '*Changes Requested*'
      } else {
        status = `*${pr.reviews.totalCount} approvals*`
      }
      text = text.concat(`\nðŸ‘‰ <${pr.url}|${pr.title}> | ${status} | ${format(pr.createdAt, 'en_US')}`)
    })

    const message = {
      blocks: [
        {
          type: "section",
          text: {
            type: "mrkdwn",
            text: `\n*${repoName}* has ${readyPRS.length} PRs ready for review`
          }
        },
        {
          type: "divider"
        },
        {
          type: "section",
          text: {
            type: "mrkdwn",
            text
          }
        },
        {
          type: "context",
          elements: [
            {
              type: "mrkdwn",
              text: `You have *${pullRequests.length}* open PRs and *${readyPRS.length}* ready for review`
            }
          ]
        }
      ],
      username: 'PR Reporter',
      icon_emoji: ':rolled_up_newspaper:' 
    }

    return await axios.post(slackWebhook, message)
  } catch (error) {
    core.setFailed(error.message)
  }
}

run()
