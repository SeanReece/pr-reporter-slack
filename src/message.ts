import { format } from 'timeago.js'
import * as core from '@actions/core'
import * as github from './github'
import { differenceInCalendarDays } from 'date-fns'
export interface BlockMessage {
  username: string
  icon_emoji: string
  blocks: object[]
}

export function formatSinglePR(pr: github.PullRequest): string {
  const stalePrDays: string = core.getInput('stale-pr') // Number of days before marking a PR as stale
  let status = ''
  let stalePr: boolean

  const createdAt = new Date(pr.createdAt)
  if (differenceInCalendarDays(createdAt, Date.now()) <= (stalePrDays || -15)) {
    stalePr = true
  }
  if (pr.reviews.totalCount === 0) {
    status = '*No reviews*'
  } else if (
    pr.reviews.nodes.some(
      review => review.state === github.ReviewStates.CHANGES_REQUESTED,
    )
  ) {
    status = '*Changes Requested*'
  } else {
    status = `*${pr.reviews.totalCount} approvals*`
  }

  const dateString = (): string => {
    if (stalePr) {
      return `ðŸš¨ ${format(pr.createdAt, 'en_US')} ðŸš¨`
    }
    return `${format(pr.createdAt, 'en_US')}`
  }

  return `\nðŸ‘‰ <${pr.url}|${pr.title}> | ${status} | ${dateString()}`
}

export function formatSlackMessage(
  repoName: string,
  text: string,
  totalPRs: number,
  readyPRs: number,
): BlockMessage {
  return {
    username: 'PR Reporter',
    // eslint-disable-next-line @typescript-eslint/camelcase
    icon_emoji: ':rolled_up_newspaper:',
    blocks: [
      {
        type: 'section',
        text: {
          type: 'mrkdwn',
          text: `\n*${repoName}* has ${readyPRs} PRs ready for review`,
        },
      },
      {
        type: 'divider',
      },
      {
        type: 'section',
        text: {
          type: 'mrkdwn',
          text,
        },
      },
      {
        type: 'context',
        elements: [
          {
            type: 'mrkdwn',
            text: `You have *${totalPRs}* open PRs and *${readyPRs}* ready for review`,
          },
        ],
      },
      {
        type: 'divider',
      },
    ],
  }
}
