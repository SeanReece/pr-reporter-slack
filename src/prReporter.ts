import * as core from '@actions/core'
import axios from 'axios'
import * as github from './github'
import { formatSlackMessage, formatSinglePR } from './message'
import { version } from '../package.json'

export default async function run(): Promise<void> {
  try {
    console.log('LOGGING')
    console.log(core)
    const token: string = core.getInput('repo-token')
    const slackWebhook: string = core.getInput('slack-webhook')
    const notifyEmpty: boolean = core.getInput('notify-empty') === 'true'
    const excludeLabels: string[] = core.getInput('exclude-labels')?.split(',')
    console.log(version)
    console.log(excludeLabels)
    core.debug(`VERSION: ${version}`)

    const response = await github.queryPRs(token)

    core.debug('Successful GraphQL response')

    const pullRequests = response?.pullRequests.nodes
    const repoName = response?.nameWithOwner

    const readyPRS = pullRequests.filter((pr: github.PullRequest) => {
      const inProgress =
        pr.isDraft || pr.title.toLowerCase().startsWith('[wip]')
      const excluded =
        excludeLabels &&
        pr.labels.nodes.some(label => excludeLabels.includes(label.name))
      return !inProgress && !excluded
    })

    let text = ''

    if (readyPRS.length === 0) {
      if (notifyEmpty) {
        text = 'üëç No PRs waiting for review!'
      } else {
        return
      }
    }

    for (const pr of readyPRS) {
      text = text.concat(formatSinglePR(pr))
    }

    const message = formatSlackMessage(
      repoName,
      text,
      pullRequests.length,
      readyPRS.length,
    )

    await axios.post(slackWebhook, message)
    core.debug('Successful Slack webhook response')
  } catch (error) {
    core.setFailed(error.message)
  }
}
