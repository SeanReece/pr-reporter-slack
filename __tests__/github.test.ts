import * as core from '@actions/core'
import { queryPRs } from '../src/github'

const mockGraphQl = jest.fn()

jest.mock('@actions/github', () => ({
  GitHub: jest.fn().mockImplementation(() => {
    return { graphql: mockGraphQl }
  }),
  context: {
    repo: 'SeanReece/pr-reporter-slack'
  }
}))
jest.mock('@actions/core')

test('performs query', async () => {
    await queryPRs('123')
    expect(mockGraphQl.mock.calls[0]).toMatchSnapshot()
})

test('throws on error', async () => {
    mockGraphQl.mockRejectedValue(new Error('bad'))
    await expect(queryPRs('123')).rejects.toThrow()
    expect(core.error).toHaveBeenCalledWith('Could not perform github query: bad')
})
