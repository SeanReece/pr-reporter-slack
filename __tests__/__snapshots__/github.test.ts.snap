// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`performs query 1`] = `
Array [
  "query prs($owner: String!, $repo: String!) {
        repository(owner:$owner, name:$repo) {
          nameWithOwner,
          pullRequests(first: 100, states: OPEN) {
            nodes {
              id
              title
              url
              createdAt
              isDraft
              reviews(first: 10, states: [CHANGES_REQUESTED, APPROVED]) {
                totalCount
                nodes {
                  state
                }
              }
              comments {
                totalCount
              }
              headRef {
                name
              }
              commits(first: 10) {
                nodes {
                  commit {
                    status {
                      id
                      state
                    }
                  }
                }
              }
            }
            totalCount
          }
        }
      }",
  Object {
    "0": "S",
    "1": "e",
    "10": "p",
    "11": "r",
    "12": "-",
    "13": "r",
    "14": "e",
    "15": "p",
    "16": "o",
    "17": "r",
    "18": "t",
    "19": "e",
    "2": "a",
    "20": "r",
    "21": "-",
    "22": "s",
    "23": "l",
    "24": "a",
    "25": "c",
    "26": "k",
    "3": "n",
    "4": "R",
    "5": "e",
    "6": "e",
    "7": "c",
    "8": "e",
    "9": "/",
    "headers": Object {
      "accept": "application/vnd.github.shadow-cat-preview+json",
    },
  },
]
`;
